LSS Simulation with `CAMB/axionCAMB` Transfer Function (#122)
hyschive/gamer-fork: tool/inits/GEN_UM_IC_WITH_PHASE_WITH_TRANSFER_FUNC/README.txt

The procedures of this LSS simulation includes:

> (a) Using CAMB/axionCAMB to produce transfer function (the input file planck_2018.ini/planck_2018_axion.ini for CAMB/axionCAMB is attached for reference)

Reviewer: How about providing the references or links for CAMB/axionCAMB in case some users may not know?

Reviewer: And remind the users that we need to install them.

Author: > How about providing the references or links for CAMB/axionCAMB in case some users may not know?

Good point! 
> And remind the users that we need to install them.

Added~

# Note
  1. CAMB will automatically compute \sigma_8 based on the given cosomology parameters. If you find the given \sigma_8 is not the desired value, it can be tuned by changing the As (scalar_amp in the input file .ini). Say if one wants to change the original \simga_8 = 0.8119112  with As = 2.100549e-9 , to new value \simga_8 = 0.818 , one just changes the new As to (0.818/0.8119112)^2*2.100549e-9 = 2.132173e-09 , then CAMB will give new \simga_8 = 0.818. This is due to power spectrum is proportional to As*f(k) .

> (b) Run the script "./set_velocities_zero.py F(T)" to produce the CAMB(axionCAMB) transfer function: planck_2018_transfer_out_no_vel.dat(planck_2018_axion_transfer_out_no_vel.dat) needed by MUSIC.

Reviewer: I cannot just run "./set_velocities_zero.py F(T)" on eureka. I have to run `python3 set_velocities_zero.py F(T)`. Also for "./make_umic_from_hdf5.py S(D) hdf5_filename".

Author: Sorry, my script has been `chmod`ed and with `!#/usr/bin/env python3.7` added at the beginning of the files, so I could execute it directly. To fix this issue, the lines `!#/usr/bin/env python3.7` have been removed, and the commands in `README.txt` have been modified to `python3 ./XXX`.

> (b) Run the script "./set_velocities_zero.py F(T)" to produce the CAMB(axionCAMB) transfer function: planck_2018_transfer_out_no_vel.dat(planck_2018_axion_transfer_out_no_vel.dat) needed by MUSIC.

Reviewer: How about mentioning that the output we want here from the previous step is `planck_2018_trnasfer_out.dat` and `planck_2018_axion_trnasfer_out.dat`?

Author: Modified~ I mention the expected output filenames in step (a) after modification.

> (c) Modifying the cosmology paramters consistent with CAMB/axionCAMB .ini files, as well as the input file parameters in ics_example.conf (transfer = camb_file; transfer_file = planck_2018_axion_transfer_out_no_vel.dat/planck_2018_axion_transfer_out_no_vel.dat; format = generic; filename = planck_2018_tf_no_vel.hdf5/planck_2018_axion_tf_no_vel.hdf5). Then run the MUSIC: "./MUSIC ics_example.conf".

Reviewer: This line seems too long. How about separating it into multiple lines?
And how about providing a reference or link for users and letting them know we need to install MUSIC first?

Author: Done~

> (c) Modifying the cosmology paramters consistent with CAMB/axionCAMB .ini files, as well as the input file parameters in ics_example.conf (transfer = camb_file; transfer_file = planck_2018_axion_transfer_out_no_vel.dat/planck_2018_axion_transfer_out_no_vel.dat; format = generic; filename = planck_2018_tf_no_vel.hdf5/planck_2018_axion_tf_no_vel.hdf5). Then run the MUSIC: "./MUSIC ics_example.conf".

Reviewer: These two are the same: `transfer_file = planck_2018_axion_transfer_out_no_vel.dat/planck_2018_axion_transfer_out_no_vel.dat`

Author: Fixed~

> (d) Run the script "./make_umic_from_hdf5.py S(D) hdf5_filename" to produce UM_IC file for single/double precision GAMER.

Reviewer: How about just using the consistent filenames following the previous step like: "./make_umic_from_hdf5.py S(D) planck_2018_tf_no_vel.hdf5"?

Author: Modified~

> (d) Run the script "./make_umic_from_hdf5.py S(D) hdf5_filename" to produce UM_IC file for single/double precision GAMER.

Reviewer: How about "for single/double precision UM_IC" since the precision of GAMER and UM_IC can be independent now?

Author: Modified~

> (e) Run the GAMER.

Reviewer: I think it will be better if the constructed UM_IC with the default parameters here can match and be directly used in the default test problem like `LSS_Hybrid` (where we need one more step to convert it to UM_IC_hybrid) such that users can easily practice the procedures without changing parameters. (For now, to run in `LLS` we also need to change `A_INIT`, `OPT__UM_IC_NVAR`, and `LSS_InitMode`.)

Reviewer: I still think changing the parameters here to align with `LSS_Hybrid`'s setup can make the workflow easier.

The procedures of this LSS simulation includes:
---------------------------------------------------------------------------------------------------------------
    Flow chart
         if not axion transfer function           if axion transfer function
           |                                           |
           | "planck_2018.ini"                         | "planck_2018_axion.ini"
           |                                           |
           v                                           v
         [(a) CAMB]                                  [(a) axionCAMB]

>          [camb planck_2018.ini]                      [./camb planck_2018_axion.ini]

Reviewer: I would like to point out that if I want to run in this directory, I have to copy not only the executable `camb` from axionCAMB, but also the file `HighLExtrapTemplate_lenspotentialCls.dat`

           |                                           |
           | "planck_2018_transfer_out.dat"            | "planck_2018_axion_transfer_out.dat"
           |                                           |
           |                                           |
           ------->  [(b) set_velocities_zero.py] <-----

>                      [python3 ./set_velocities_zero.py F planck_2018_trnasfer_out.dat planck_2018_transfer_out_no_vel.dat (CAMB)]
>                      [python3 ./set_velocities_zero.py T planck_2018_axion_trnasfer_out.dat planck_2018_axion_transfer_out_no_vel.dat (axionCAMB)]

Reviewer: ```suggestion
                     [python3 ./set_velocities_zero.py F planck_2018_transfer_out.dat planck_2018_transfer_out_no_vel.dat (CAMB)]
                     [python3 ./set_velocities_zero.py T planck_2018_axion_transfer_out.dat planck_2018_axion_transfer_out_no_vel.dat (axionCAMB)]
```

                                 |
                                 | "planck_2018_transfer_out_no_vel.dat" (CAMB)
                                 | "planck_2018_axion_transfer_out_no_vel.dat" (axionCAMB)
                                 |
       "ics_example.conf"        v
      ----------------------> [(c) MUSIC]
                              [./MUSIC ics_example.conf]
                                 |
                                 | "planck_2018_tf_no_vel.hdf5" (CAMB)
                                 | "planck_2018_axion_tf_no_vel.hdf5" (axionCAMB)
                                 |
      "ics_example.conf"         v
      --------------> [(d) make_umic_from_hdf5.py]

>                       [python3 ./make_umic_from_hdf5.py S(D) planck_2018_tf_no_vel.hdf (CAMB)]
>                       [python3 ./make_umic_from_hdf5.py S(D) planck_2018_axion_tf_no_vel.hdf (axionCAMB)]

Reviewer: ```suggestion
                      [python3 ./make_umic_from_hdf5.py S(D) planck_2018_tf_no_vel.hdf5 (CAMB)]
                      [python3 ./make_umic_from_hdf5.py S(D) planck_2018_axion_tf_no_vel.hdf5 (axionCAMB)]
```

                                 |
                                 | "UM_IC"
                                 |
                                 v
                              [(e) GAMER]

---------------------------------------------------------------------------------------------------------------

> (a) After successful installation of CAMB/axionCAMB (see Notes for installation), use CAMB/axionCAMB to to produce transfer function.

Reviewer: ```suggestion
(a) After successful installation of CAMB/axionCAMB (see Notes for installation), use CAMB/axionCAMB to produce the transfer function.
```

    The input file planck_2018.ini/planck_2018_axion.ini for CAMB/axionCAMB is attached for reference.
    The expected output files are:

>         planck_2018_trnasfer_out.dat       -> for CAMB

Reviewer: ```suggestion
        planck_2018_transfer_out.dat       -> for CAMB
```

        planck_2018_axion_trnasfer_out.dat -> for axionCAMB

(b) Run the script: "python3 ./set_velocities_zero.py F(T) planck_2018_trnasfer_out.dat(planck_2018_axion_trnasfer_out.dat) planck_2018_transfer_out_no_vel.dat(planck_2018_axion_transfer_out_no_vel.dat)".

>     The script reads-in the output files in step(a), and produce the transfer function requireded by MUSIC.

Reviewer: ```suggestion
    The script reads in the output files in step(a), and produces the transfer function required by MUSIC.
```

    The boolean F/T suggests whether axion-physics is included in the input transfer function files.
    The expected output files are:
        planck_2018_transfer_out_no_vel.dat       -> for CAMB
        planck_2018_axion_transfer_out_no_vel.dat -> for axionCAMB

> (c) Modifying the cosmology paramters to be consistent with CAMB/axionCAMB .ini files, as well as the input file parameters in MUSIC input file ics_example.conf:

Reviewer: ```suggestion
(c) Modify the cosmology parameters to be consistent with CAMB/axionCAMB .ini files, as well as the input file parameters in the MUSIC input file ics_example.conf:
```

        transfer      = camb_file
        transfer_file = planck_2018_transfer_out_no_vel.dat/planck_2018_axion_transfer_out_no_vel.dat
        format        = generic
        filename      = planck_2018_tf_no_vel.hdf5/planck_2018_axion_tf_no_vel.hdf5)
    Install (see Notes for installation) and run the MUSIC: "./MUSIC ics_example.conf".

(d) Run the script "python3 ./make_umic_from_hdf5.py S(D) planck_2018_axion_tf_no_vel.hdf5" to produce UM_IC file for single(double) precision UM_IC for FDM simulation (or "python3 ./make_umic_from_hdf5.py S(D) planck_2018_tf_no_vel.hdf5" for CDM simulation).

(e) Run the GAMER.

# Notes
  1. CAMB will automatically compute \sigma_8 based on the given cosomology parameters.
     If you find the given \sigma_8 is not the desired value, it can be tuned by changing the As (scalar_amp in the input file .ini).
     Say if one wants to change the original \sigma_8 = 0.8119112  with As = 2.100549e-9 , to new value \sigma_8 = 0.818 , one just changes the new As to (0.818/0.8119112)^2*2.100549e-9 = 2.132173e-09 , then CAMB will give new \sigma_8 = 0.818.

>      This is due to power spectrum is proportional to As*f(k) .

Reviewer: ```suggestion
     This is because the power spectrum is proportional to As*f(k).
```

  2. CAMB (Code for Anisotropies in the Microwave Background)
         GitHub page:                    https://github.com/cmbant/CAMB

>          Install procedures:             https://github.com/cmbant/CAMB#description-and-installation
>   3. axionCAMB(modified version of CAMB to include axion physics)
>          GitHub page:                    https://github.com/dgrin1/axionCAMB
>          Install procedures:             https://github.com/dgrin1/axionCAMB?tab=readme-ov-file#compiling-axioncamb
>   4. MUSIC(MUlti-Scale Initial Conditions for cosmological simulations)
>          Bitbucket page:                 https://bitbucket.org/ohahn/music/src/master/
>          Website (for user's guide pdf): https://www-n.oca.eu/ohahn/MUSIC/
>          Install procedures:             modified the path and parameters in MUSIC-repo-provided Makefile, then make

Reviewer: ```suggestion
         Installation procedures:        https://github.com/cmbant/CAMB#description-and-installation
  3. axionCAMB(modified version of CAMB to include axion physics)
         GitHub page:                    https://github.com/dgrin1/axionCAMB
         Installation procedures:        https://github.com/dgrin1/axionCAMB?tab=readme-ov-file#compiling-axioncamb
  4. MUSIC(MUlti-Scale Initial Conditions for cosmological simulations)
         Bitbucket page:                 https://bitbucket.org/ohahn/music/src/master/
         Website (for user's guide pdf): https://www-n.oca.eu/ohahn/MUSIC/
         Installation procedures:        modified the path and parameters in MUSIC-repo-provided Makefile, then make
```
